cmd/bot/run.go

```diff
@@ -0,0 +1,305 @@
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/indexer/config"
+	oplog "github.com/ethereum-optimism/optimism/op-service/log"
+	"github.com/ethereum-optimism/optimism/op-service/retry"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/urfave/cli/v2"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+func runCommand(ctx *cli.Context) error {
+	// Config logger
+	logger := oplog.NewLogger(oplog.AppOut(ctx), oplog.ReadCLIConfig(ctx)).New("role", "bot")
+	oplog.SetGlobalLogHandler(logger.GetHandler())
+	logger.Info("opbnb-bridge-bot is starting")
+
+	// Load config
+	cfg, err := core.LoadConfig(logger, ctx.String(ConfigFlag.Name))
+	if err != nil {
+		logger.Error("failed to load config", "err", err)
+		return err
+	}
+
+	// Connect to L1 and L2 RPCs
+	l1Client, err := core.Dial(cfg.RPCs.L1RPC)
+	if err != nil {
+		return fmt.Errorf("dial endpoint %s: %w", cfg.RPCs.L1RPC, err)
+	}
+	l2Client, err := core.Dial(cfg.RPCs.L2RPC)
+	if err != nil {
+		return fmt.Errorf("dial endpoint %s: %w", cfg.RPCs.L2RPC, err)
+	}
+
+	// Connect to database and ensure schemas initialized
+	db, err := connect(logger, cfg.DB)
+	if err != nil {
+		return fmt.Errorf("failed to connect database: %w", err)
+	}
+	err = db.AutoMigrate(&core.L2ScannedBlock{})
+	if err != nil {
+		return fmt.Errorf("failed to migrate l2_scanned_blocks: %w", err)
+	}
+	err = db.AutoMigrate(&core.L2ContractEvent{})
+	if err != nil {
+		return fmt.Errorf("failed to migrate l2_contract_events: %w", err)
+	}
+
+	l2ScannedBlock, err := queryL2ScannedBlock(db, &cfg)
+	if err != nil {
+		return err
+	}
+
+	go WatchBotDelegatedWithdrawals(ctx.Context, logger, db, l2Client, l2ScannedBlock, cfg)
+	go ProcessBotDelegatedWithdrawals(ctx.Context, logger, db, l1Client, l2Client, cfg)
+
+	<-ctx.Context.Done()
+	return nil
+}
+
+// ProcessBotDelegatedWithdrawals processes the indexed bot-delegated withdrawals.
+// It will prove the withdrawal transaction when the proposal time window has passed;
+// and it will finalize the withdrawal when the challenge time window has passed.
+func ProcessBotDelegatedWithdrawals(ctx context.Context, log log.Logger, db *gorm.DB, l1Client *core.ClientExt, l2Client *core.ClientExt, cfg core.Config) {
+	ticker := time.NewTicker(3 * time.Second)
+	processor := core.NewProcessor(log, l1Client, l2Client, cfg)
+	limit := 1000
+
+	for {
+		select {
+		case <-ticker.C:
+		case <-ctx.Done():
+			return
+		}
+
+		unprovens := make([]core.L2ContractEvent, 0)
+		if result := db.Order("id asc").Where("proven = false").Limit(limit).Find(&unprovens); result.Error != nil {
+			log.Error("querying l2_contract_events", "error", result.Error)
+			continue
+		} else if len(unprovens) > 0 {
+			for _, unproven := range unprovens {
+				if unproven.BlockTime+cfg.Misc.ProposeTimeWindow < time.Now().Unix() {
```

Comment:
Normally too deep nested code is hard to read. I suggest rewrite it by handle earlier with `esle` condition.

E.g. 

```
if unproven.BlockTime+cfg.Misc.ProposeTimeWindow >= time.Now().Unix() {
    continue
}
original_if_code_block()
```

Reply:
Another PR has been created to refactor this function, to make it more readable. PTAL.